File: RunningProduct.py
# Time Complexity: O(N) +O(M) where O(N) time taken to traverse left side of array and O(M) is time taken to traverse right side of array
# Space Complexity: O(1) as Output array result is not considered as extra space in single array calculation is done
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        result = []
        result.append(1)
        main = 1
        i = 1
        while i < len(nums):
            main = nums[i-1]*main
            result.append(main)
            i+=1
        j = len(nums)-2
        main = 1
        while j>=0:
            main = nums[j+1]*main
            result[j] = result[j]*main
            j-=1
        return result
        
            
        

File: RunningProduct.java
// Time Complexity : O(N)
// Auxilary space O(1)
// Did this code successfully run on Leetcode : Yes

class Solution {
    public int[] productExceptSelf(int[] nums) {
        if(nums == null || nums.length==0){
            return new int[] {};
        }
        int result[] = new int[nums.length];
        int rp = 1;
        for(int i=0;i<nums.length;i++){
            result[i] = rp;
            rp = rp * nums[i];
        }
        
        rp = 1;
        for(int i=nums.length-1;i>=0;i--){
            result[i] = result[i] * rp;
            rp = rp * nums[i];
        }
        return result;
    }
}